### PR TITLE
Add mod tags

### PR DESCRIPTION
**Pull request by Feyko**

### CODE REVIEW

File: db/postgres/modtags.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+func CreateModTag(modTag *ModTag, ctx *context.Context) (*ModTag, error) {
+	if len(modTag.Name) > 20 {
+		return nil, errors.New("Mod tag name is more than 20 characters")
+	}
+
+	exists := GetModTagByName(modTag.Name, ctx)
+	if exists != nil {
+		return nil, errors.New("Mod tag already exists")
+	}
```
Comment (Vilsol) [inline db/postgres/modtags.go:18]: You don't need to check if it exists, try inserting it and see if you got an error in response

File: gql/resolver_modtags.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	modTag := postgres.GetModTagByName(modTagName, &newCtx)
+
+	return DBModTagToGenerated(modTag), nil
+}
+
+func (r *queryResolver) GetModTags(ctx context.Context, filter *generated.ModTagFilter) ([]*generated.ModTag, error) {
+	wrapper, newCtx := WrapQueryTrace(ctx, "getModTags")
+	defer wrapper.end()
+
+	modTags := postgres.GetModTags(&newCtx)
```
Comment (Vilsol) [inline gql/resolver_modtags.go:73]: Filter is not passed to DB

**Review by Vilsol — CHANGES_REQUESTED** [review]

File: db/postgres/modtags.go (hunk 1 above)
Comment (Feyko) [reply db/postgres/modtags.go:18]: Got it

File: gql/resolver_modtags.go (hunk 2 above)
Comment (Feyko) [reply gql/resolver_modtags.go:73]: Yep that's just a wip so that generation passes

File: db/postgres/postgres_types.go [hunk 3]
```diff
@@ -155,3 +158,25 @@ type Announcement struct {
...
+
+	Name string `gorm:"primary_key;type:varchar(20)"`
+
+	Mods []Mod `gorm:"many2many:mod_tags"`
+}
+
+type ModTag struct {
+	SMRModel
+
+	TagID string `gorm:"primary_key;type:varchar(20)"`
```
Comment (Vilsol) [inline db/postgres/postgres_types.go:173]: ID is 16 characters

File: db/postgres/postgres_types.go [hunk 4]
```diff
@@ -155,3 +158,25 @@ type Announcement struct {
...
+	ModID string `gorm:"primary_key;type:varchar(16)"`
+}
+
+type GuideTag struct {
+	SMRModel
+
+	TagID   string `gorm:"primary_key;type:varchar(20)"`
```
Comment (Vilsol) [inline db/postgres/postgres_types.go:180]: ID is 16 characters

File: db/postgres/tags.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	if len(tag) > 20 {
+		return errors.New("Tag name is over 20 characters long")
+	}
+	if moderation.IsInappropriate(tag) {
+		return errors.New("Tag name is inapproriate")
+	}
+	return nil
+}
+
+func CreateTag(tag *Tag, ctx *context.Context, ratelimit bool) (*Tag, error) {
```
Comment (Vilsol) [inline db/postgres/tags.go:25]: Unused `ratelimit` argument

File: db/postgres/tags.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+	err := ValidateTagName(tag.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	exists := GetTagByName(tag.Name, ctx)
```
Comment (Vilsol) [inline db/postgres/tags.go:31]: Should attempt to insert and check for response error if fails, no need to check before creating.

File: generated/custom_models.go [hunk 7]
```diff
@@ -10,6 +10,7 @@ type NewMod struct {
 	SourceURL        *string         `json:"source_url"`
 	ModReference     string          `json:"mod_reference"`
 	Hidden           *bool           `json:"hidden"`
+	Tags             []string        `json:"tags"`
```
Comment (Vilsol) [inline generated/custom_models.go:13]: Add validation

File: generated/custom_models.go [hunk 8]
```diff
@@ -21,4 +22,5 @@ type UpdateMod struct {
 	ModReference     *string         `json:"mod_reference"`
 	Hidden           *bool           `json:"hidden"`
 	Authors          []UpdateUserMod `json:"authors"`
+	Tags             []string        `json:"tags"`
```
Comment (Vilsol) [inline generated/custom_models.go:25]: Add validation

File: db/postgres/tags.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+func ValidateTagName(tag string) error {
+	if len(tag) > 20 {
+		return errors.New("Tag name is over 20 characters long")
+	}
+	if moderation.IsInappropriate(tag) {
+		return errors.New("Tag name is inapproriate")
+	}
+	return nil
+}
```
Comment (Vilsol) [inline db/postgres/tags.go:23]: This should be handled before it reaches the database

File: schemas/mod.graphql [hunk 10]
```diff
@@ -13,6 +13,7 @@ enum ModFields {
     popularity
     last_version_date
     search
+    tags
```
Comment (Vilsol) [inline schemas/mod.graphql:16]: This is only used for fields that can be ordered by. A list of tags cannot be ordered by

File: db/postgres/postgres_types.go [hunk 11]
```diff
@@ -155,3 +158,25 @@ type Announcement struct {
...
+
+type ModTag struct {
+	SMRModel
+
+	TagID string `gorm:"primary_key;type:varchar(20)"`
+	ModID string `gorm:"primary_key;type:varchar(16)"`
+}
+
+type GuideTag struct {
+	SMRModel
```
Comment (Vilsol) [inline db/postgres/postgres_types.go:178]: You don't need `SMRModel`, but rather `SMRDates` as `SMRModel` includes an `id` primary key field

File: db/postgres/postgres_types.go [hunk 12]
```diff
@@ -155,3 +158,25 @@ type Announcement struct {
...
+type Tag struct {
+	SMRModel
+
+	Name string `gorm:"primary_key;type:varchar(20)"`
+
+	Mods []Mod `gorm:"many2many:mod_tags"`
+}
+
+type ModTag struct {
+	SMRModel
```
Comment (Vilsol) [inline db/postgres/postgres_types.go:171]: You don't need `SMRModel`, but rather `SMRDates` as `SMRModel` includes an `id` primary key field

**Review by Vilsol — CHANGES_REQUESTED** [review]

**Pull request closed without merging**